File: my-folder/0733-flood-fill/solution.py
class Solution:
    def floodFill(self, image: List[List[int]], sr: int, sc: int, color: int) -> List[List[int]]:
        def dfs(image, row, col, oriCol, color):
            if row < 0 or row >= len(image) or col < 0 or col >= len(image[0]) or image[row][col] != oriCol:
                return
            image[row][col] = color
            dfs(image, row - 1, col, oriCol, color)
            dfs(image, row + 1, col, oriCol, color)
            dfs(image, row, col - 1, oriCol, color)
            dfs(image, row, col + 1, oriCol, color)

        originalColor = image[sr][sc]
        if originalColor != color:
            dfs(image, sr, sc, originalColor, color)
        
        return image

File: my-folder/0733-flood-fill/solution.java
class Solution {
    public int[][] floodFill(int[][] image, int sr, int sc, int color) {
        int theColor = image[sr][sc];
        if (theColor != color) {
            dfs(image, sr, sc, theColor, color);
        }
        return image;
    }

    public void dfs (int[][] image, int r, int c, int color, int newColor) {
        if (image[r][c] == color) {
            image[r][c] = newColor;
            if (r >= 1) {
                dfs (image, r-1, c, color, newColor);
            }
            if (c >= 1) {
                dfs (image, r, c-1, color, newColor);
            }
            if (r+1 < image.length) {
                dfs (image, r+1, c, color, newColor);
            }
            if (c+1 < image[0].length) {
                dfs (image, r, c+1, color, newColor);
            }
        }
    }
}
